File: Backend/routes/book_in_bookshelf.py
from fastapi import APIRouter
from database.crud_book_in_bookshelf import insert_book_in_bookshelf_db, get_book_in_bookshelf_db, update_book_in_bookshelf_db


book_in_bookshelf = APIRouter()


@book_in_bookshelf.post('/')
def insert_book_in_bookshelf(id_book: str, id_bookshelf: str):
    return insert_book_in_bookshelf_db(id_book, id_bookshelf)

@book_in_bookshelf.get('/')
def get_book_in_bookshelf(id_bookshelf: str = None, id_book: str = None, id_user: str = None):
    return get_book_in_bookshelf_db(id_bookshelf, id_book,  id_user)

@book_in_bookshelf.put('/')
def update_book_in_bookshelf(last_id_bookshelf: str, new_id_bookshelf: str, id_book: str):
    return update_book_in_bookshelf_db(last_id_bookshelf, new_id_bookshelf, id_book)

File: Frontend/digital-library/src/action/API/bookInBookshelf.js
const URL_API = 'http://127.0.0.1:8000/';

export const insertBookInBookshelf = (idBook, idBookshelf) => {
    return new Promise((resolve, reject) => {
        fetch(`${URL_API}book-in-bookshelf/?id_book=${idBook}&id_bookshelf=${idBookshelf}`, { method: 'POST', headers: { 'Content-Type':  'application/json' }, })
            .then(response =>  response.json())
            .then(data => resolve(data))
            .catch(error => reject(error));
    })
}

export const getBookInBookshelf = (idBook, idUser) => {
    return new Promise((resolve, reject) => {
        fetch(`${URL_API}book-in-bookshelf/?id_book=${idBook}&id_user=${idUser}`)
            .then(response => response.json())
            .then(data => resolve(data))
            .catch(error => reject('Error inserting bookshelf', error));
    })
}

export const getBookInBookshelfByIdBookshelf = (idBookshelf) => {
    return new Promise((resolve, reject) => {
        fetch(`${URL_API}book-in-bookshelf/?id_bookshelf=${idBookshelf}`)
            .then(response => response.json())
            .then(data => resolve(data))
            .catch(error => reject('Error inserting bookshelf', error));
    })
}

export const updateBookInBookshelf = (lastId, newId, idBook) => {
    return new Promise((resolve, reject) => {
        fetch(`${URL_API}book-in-bookshelf/?last_id_bookshelf=${lastId}&new_id_bookshelf=${newId}&id_book=${idBook}`, {
              method: 'PUT',
              headers: {
                'Content-Type': 'application/json'
              },
            })
            .then(response => response.json())
            .then(data => resolve(data))
            .catch(error => reject('Error inserting bookshelf', error));
    })
}

export  const deleteBookInBookshelf = (idBook, idBookshelf) => {   
    return new Promise((resolve, reject) => {
        fetch(`${URL_API}book-in-bookshelf/?id_book=${idBook}&id_bookshelf=${idBookshelf}`, { 
            method: 'DELETE', 
            headers: { 
                'Content-Type': 'application/json' 
            }, 
        })
        .then(response => { return response.json(); })
        .then(data => resolve(data))
        .catch(error => reject('Error inserting bookshelf', error));
    })
}

export const getBookInBookshelfByIdBook = (idBook, idUser) => {
    return new Promise((resolve, reject) => {
        fetch(`${URL_API}book-in-bookshelf/?id_book=${idBook}&id_user=${idUser}`)
            .then(response => response.json())
            .then(data => resolve(data))
            .catch(error => reject('Error inserting bookshelf', error));
    })
}